### ISSUE TITLE
Shade Types

### ISSUE THREAD

**Issue by imrichbiotch** [schemers/homebridge-hunterdouglas-platinum#6]
Hi @schemers @twangus,

Wondering if you guys could take a shot at seeing if the "Current Horizontal [Vertical] Tilt Angle" and "Target Horizontal [Vertical] Tilt Angle" is data that can be pulled from the Platinum controller.  On the iOS app, it appears as "Vanes Open/Close"

![IMG_E58DD6DA0196-1](https://user-images.githubusercontent.com/65510811/82284943-16ab1b80-994f-11ea-921f-42615ce36471.jpeg)

**Comment by twangus** [comment]
Yeah, this should be possible, but it might depend on the specific type of shade that you have.  

My understanding is that the Platinum hub uses ID numbers corresponding to the different features of the various shade models.  Feature ID "4" corresponds to the basic bottom up raise/lower feature in most shades.  On top-down / bottom-up shades (like I have), ID "18" corresponds to the bottom-up feature.  The platinum hub uses those IDs to report out the shade's current position, and we can use them to set the value for that feature.

From other code I've seen, the hub uses feature ID "7" for the vane position on Silhouette and Nantucket shades.  So for those types of shades, the plugin should be able to read and control the bottom-up feature with ID "4" and the vane feature with ID "7."  But I don't have those shades and haven't confirmed that.

One potentially weird thing is that it seems like the hub uses ID "7" for different features on other types of shades.  Supposedly, some Luminette products use ID "4" for a "traverse" feature, ID "7" for a "close right" feature, and ID "19" for a "close left" feature.  Again, I haven't confirmed this and don't have those shades so I'm not even really sure what they correspond to in practice.

Anyway, in summary, it should be possible for the plugin to read and control the feature but I can't say for sure and it might depend on the shades.

The bigger question is how the plugin should present the vane feature to HomeKit.  HomeKit accessories can have multiple "services," each of which can have multiple "characteristics."  I believe that, currently, each accessory created by the plugin uses a single Window Covering service with Current Position and Target Position characteristics.  In the original plugin, those characteristics service always controlled the bottom-up functionality (ID 4).  In the code I added for preliminary support of top-down / bottom-up shades, the user can choose whether to have the service control top-down (ID 18) or bottom-up (ID 4) functionality, but it doesn't yet provide a way to control both.  That was still helpful for me, since I almost always use the top-down feature, but it probably wouldn't make sense for the vane feature, where I assume you wouldn't want the vane control to come at the expense of the standard open/close feature.

For more sophisticated top-down / bottom-up support, I was thinking about updating the plugin to provide multiple "services" for a single blind, which HomeKit allows.  That way, the UI would provide sliders for both "top-down" and a "bottom-up" services and the plugin could translate those correct feature IDs in the platinum hub (i.e., 4 and 18, respectively).  I created a test accessory using the HomeKit Accessory Simulator, and the Home app provides nice side-by-side sliders for each service.

![TDBU](https://user-images.githubusercontent.com/31568/82702612-2a909f00-9c27-11ea-8fcb-dcccaabc33d6.jpeg)

For vane support, you shouldn't even need to create a separate "service," because HomeKit provides specific "characteristics" for reading and controlling the tilt angle of a slat -- e.g., [HMCharacteristicTypeCurrentTilt](https://developer.apple.com/documentation/homekit/hmcharacteristictypecurrenttilt).  For those characteristics, a 0 value in the middle corresponds to open slats, whereas negative or positive values correspond to slats that are angled in a given direction.

Oddly, though, it doesn't seem like the Home app actually provides an interface for reading or controlling that characteristic even if the HomeKit accessory provides it.  Or, at least, I can't figure out how to get it to show such an interface.  I created a test accessory in the HomeKit Accessory Simulator and specifically added the different "tilt" characteristics, but the Home app just ignores them and only shows a single slider for the standard "position" characteristic.  I also couldn't figure out a way to get Siri to report or control the characteristic.  However, the Eve app does see that characteristic and lets you edit it, so it seems like it is set up correctly:

![TiltAngle](https://user-images.githubusercontent.com/31568/82703982-7f351980-9c29-11ea-8719-0f82d502d781.jpeg)

So, even if support for the Hunter Douglas "vane" feature was added to this plugin, I'm not sure how it should be implemented.  It seems like the "correct" way to implement it in HomeKit would be to use the "tilt angle" characteristics that are explicitly meant to describe that, but it might not be useful to anyone if the characteristic can't be read or controlled via Siri or the Home app and instead requires a third-party HomeKit app like Eve.  Alternatively, we could just ignore the "tilt angle" characteristic entirely and provide vane support via an entirely separate service using the default "position" characteristics.  That's basically what I was planning to implement someday for top-down / bottom-up shades.  But if we did that for "vane" support, the control would always show up as a slider from 0% (closed) to 100% (open), which might not make sense for vanes where the "open" position is in the middle... 

Anyway, it is worth thinking about.  I'm tied up for the next few weeks, but I might take a stab at adding support for accessories with multiple services sometime in June.  If so, I can try to see if the same approach would work for vanes.  In the meantime, you might want to download the HomeKit Accessory Simulator yourself and think about what would make the most sense for controlling those features.

**Comment by imrichbiotch** [comment]
@twangus thanks for the reply. it looks like it could potentially show up as in the Settings for that window as a slider for vanes aka Tilt Angle?

![IMG_A5D020B78870-1](https://user-images.githubusercontent.com/65510811/82721940-183e5180-9c77-11ea-84a4-2e2f832e686f.jpeg)

**Comment by schemers** [comment]
yeah, it sounds like tilt angle would work. You just have to pull up to see the angle(s).

to determine what commands get sent to bridge when you change the title angles, you can do the following in a macOS terminal window:

```
$ nc 192.168.86.143 522
```

replace the IP address with the one  for your bridge.

you should see:
```
1 HunterDouglas Shade Controller
```

then, in the platinum app open/close the blinds, and change the tilt angles. You should see more output as you do the, like:

```
1 $act01-00-
1 $cp05-04-247-
1 $act02-00-
1 $act00-00-
```

You'd want to cycle through all the various mins/max and document the exact set of commands for each one. I'd probably do:

- blind closed
- blind open
- vanes closed all the way to the left
- vanes midway between left and middle
- vanes closed in the middle (or where that middle vertical line is)
- vanes midway between middle right 
- vanes closed all the way to the right

It would also help to type (and hit return):

```
$dat; 
```

and capture that output, since it will have the blind type and other useful info it.

If you can capture that then maybe one of us can poke around when we get some time.

**Comment by imrichbiotch** [comment]
This is a lot of interesting data!

So it appears that all the horizontal vane Silhouette models function the same for me (whether or not they have a blackout shade). The blackout shade seems to appear to the system a completely different unit all together despite being a part of the same hardware for each window.

The horizontal Silhouette shades have vanes that remain visibly "closed" throughout the different shade positions (open/close/partial open). When it is fully closed or covering the window at the very bottom position, the vanes can then tilt. If I command the horizontal Silhouette shades to "open" from a vanes tilted position, the vanes shut first then the shade opens.

Here are two examples that I captured. I took these two as one is a smaller window and the other is a big sliding door. They both seem to behave similarly. You can ignore if the data doesn't appear to perfect match my 50% partial estimate as the slider on the Hunter Douglas app is not very precise.

> ### **Type 1: Horizontal Silhouette — Bedside Window**
> **Full Open**
> 1 $cp00-04-255-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Open to 50% Close**
> 1 $act01-00-
> 1 $cp00-04-124-
> 1 $act02-00-
> 1 $act00-00-
> 
> **50% Close to Close [aka Vanes Closed]**
> 1 $act01-00-
> 1 $cp00-04-000-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Vanes Close to Vanes 50%**
> 1 $act01-00-
> 1 $cp00-07-112-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Vanes 50% to Vanes Open**
> 1 $act01-00-
> 1 $cp00-07-255-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Vanes Open to Vanes 50%**
> 1 $act01-00-
> 1 $cp00-07-127-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Vanes 50% to Open Shade** 
> 1 $act01-00-
> 1 $cp00-04-255-
> 1 $act02-00-
> 1 $act00-00-
> 
> ———
> ### **Type 1: Horizontal Silhouette — Sliding Door (Larger but same function)**
> **Full Open**
> 1 $act01-00-
> 1 $cp01-04-255-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Full Open to Vanes 30%**
>  $act01-00-
> 1 $cp01-07-086-
> 1 $act02-00-
> 1 $act00-00-
> 
> **Vanes 30% to Full Open**
> 1 $act01-00-
> 1 $cp01-04-255-
> 1 $act02-00-
> 1 $act00-00-

The vertical Silhouette shade that I have is for a pair of patio sliding doors. The shade opens by parting from the center. Unlike the horizontal Silhouette shade, the vanes on this shade are in the "Vanes Open" position throughout the different shade positions (open/close/partial open). When the two halves of the shades meet in the center and the shade is considered "closed," the vanes can then tilt either left or right (i.e. CW or CCW rotation of the vertical vanes). Hope that makes sense!

> ### **Type 2: Vertical Silhouette — Patio Door (Vertical Vanes that Part in the middle to open, when both sides meet in the middle vanes can tilt left or right in unison)**
> **Full Open to 50% Close**
> 1 $act01-00-
> 1 $cp06-04-136-
> 1 $act02-02-
> 1 $act00-00-
> 
> **50% Close to Full Close [aka Vanes Open]**
> 1 $act01-00-
> 1 $cp06-04-000-
> 1 $act02-02-
> 1 $act00-00-
> 
> **Full Close/Vanes Open to Vanes 50% Close Tilt Left**
> 1 $act01-00-
> 1 $cp06-19-096-
> 1 $act02-02-
> 1 $act00-00-
> 
> **Vanes 50% Close Tilt Left to Vanes 50% Close Tilt Right**
> 1 $act01-00-
> 1 $cp06-07-078-
> 1 $act02-02-
> 1 $act00-00-
> 
> **Vanes 50% Close Tilt Right to Vanes 100% Close Tilt Right**
> 1 $act01-00-
> 1 $cp06-07-000-
> 1 $act02-02-
> 1 $act00-00-
> 
> **Vanes 100% Close Tilt Right to Vanes Open**
> 1 $act01-00-
> 1 $cp06-07-251-
> 1 $act02-02-
> 1 $act00-00-
> 
> **Vanes Open to Vanes 100% Close Tilt Left**
> 1 $act01-00-
> 1 $cp06-19-000-
> 1 $act02-02-
> 1 $act00-00-
> 
> **Vanes 100% Close Tilt Left to Open Shades**
`
1 $act01-00-
1 $cp06-04-255-
1 $act02-02-
1 $act00-00-
`

**Comment by schemers** [comment]
thanks. I removed all but the `$cp` lines, which are reporting the shade id/feature/position.

The format is:

`$cp{shade-id}-{feature}-{position-0-to-255}`

The part I don't fully understand yet is the relation between shade type and shade feature, and determining the current state of a shade since it can have multiple features, and it looks like the position for only a single feature is returned for a given shade.

I think the key is certain features can only be used when other features are in a given state. Like you said, the vanes can only tilt (feature 7 and 19) can only tilt when the blind is closed (feature 4), so maybe if the tilt feature position is being reported that implicitly means the blind is closed. 

What would help is the output of typing `$dat;` after connecting with the `nc` command, so we can see the values for `$cr/$cs/$cp`, which are used to define:

- rooms
- shades (shade type and what room they are in
- current shade position/state 

For example, I have seven shades in my master bedroom which look like:

```
1 $cr00-01-0xE6AA-Master
1 $cs00-00-04-Window Left
1 $cp00-04-255-
1 $cs01-00-02-Left Door 
1 $cp01-04-255-
1 $cs02-00-08-Door
1 $cp02-04-255-
1 $cs03-00-01-Window Right Of Door
1 $cp03-04-255-
1 $cs04-00-05-Center Bed
1 $cp04-04-255-
1 $cs05-00-06-Right Bed
1 $cp05-04-255-
1 $cs06-00-03-Left Bed
1 $cp06-04-255-
```

**Comment by schemers** [comment]
> One potentially weird thing is that it seems like the hub uses ID "7" for different features on other types of shades. Supposedly, some Luminette products use ID "4" for a "traverse" feature, ID "7" for a "close right" feature, and ID "19" for a "close left" feature. Again, I haven't confirmed this and don't have those shades so I'm not even really sure what they correspond to in practice.
> 

@twangus thanks for the detailed info. I think mapping from blind type to feature is captured in that [zip file](https://github.com/schemers/homebridge-hunterdouglas-platinum/files/4611628/PlatinumVera.zip) you shared. It shows:

- blind type 6 and 8  feature 19 is `Close Left`, and feature 7 is `Close Right`
- blind type 10 feature 7 means `Vane`

```
-- This is a list of shade types indexed by shade type + 1
-- name - The shade type name
-- order - an array of feature numbers supported by this shade type in the order in which they should be used
-- xover (optional - if shade has more than 1 feature) - The crossover point as a percentage value. Below this value, the second feature is used.
-- tdbu (optional) - Indicates special handling for top-down/bottom-up shade types
-- feature - an object matching the feature numbers to names
-- The last index (all) includes features common to all shades.
platinum_shade_types = {
    [ 0+1]={name="Alustra Woven Textures: Roller",                                   order={4},                                feature={[4]="Bottom-Up"}},
    [ 1+1]={name="Duette & Applause Honeycomb Shades: Standard",                     order={4},                                feature={[4]="Bottom-Up"}},  
    [ 2+1]={name="Duette & Applause Honeycomb Shades: DuoLite & Top-Down/Bottom-Up", order={4, 18},    xover=50, tdbu=true,    feature={[4]="Bottom Rail", [18]="Middle Rail"}},  
    [ 3+1]={name="Duette & Applause Honeycomb Shades: Top-Down",                     order={4},                                feature={[4]="Top-Down"}},  
    [ 4+1]={name="Duette & Applause Honeycomb Shades: Skylift",                      order={4},                                feature={[4]="Bottom-Up"}},  
    [ 5+1]={name="Desgner Roller & Screen Shades",                                   order={4},                                feature={[4]="Bottom-Up"}},  
    [ 6+1]={name="Luminette Privacy Sheers",                                         order={4, 19, 7}, xover=20,               feature={[4]="Traverse",    [19]="Close left",    [7]="Close right"}},
    [ 7+1]={name="Luminette Modern Draperies: Full Panel",                           order={4},                                feature={[4]="Traverse"}},  
    [ 8+1]={name="Luminette Modern Draperies: Dual Panel",                           order={4, 19, 7}, xover=20,               feature={[4]="Traverse",    [19]="Close left",    [7]="Close right"}},
    [ 9+1]={name="Pirouette Window Shadings",                                        order={4},                                feature={[4]="Bottom-Up"}},  
    [10+1]={name="Silhouette & Nanticket Window Shadings",                           order={4, 7},     xover=20,               feature={[4]="Bottom-Up",   [ 7]="Vane"}},  
    [11+1]={name="Vignette Modern Roman Shades: Traditional",                        order={4},                                feature={[4]="Bottom-Up"}},  
    [12+1]={name="Vignette Modern Roman Shades: Tiered",                             order={4},                                feature={[4]="Bottom-Up"}},  
    [13+1]={name="Vignette Modern Roman Shades: Tiered Top-Down/Bottom-Up",          order={4, 18},    xover=50, tdbu=true,    feature={[4]="Bottom Rail", [18]="Middle Rail"}},  
    [14+1]={name="Skyline Gliding Window Panels",                                    order={4},                                feature={[4]="Traverse"}},  
    [15+1]={name="Pleated Shades",                                                   order={4},                                feature={[4]="Bottom-Up"}},  
    [16+1]={name="Alustra Woven Textures: Roman",                                    order={4},                                feature={[4]="Bottom-Up"}},
    [17+1]={name="Solera Soft Shades",                                               order={4},                                feature={[4]="Bottom-Up"}},
    [18+1]={name="Design Studio Roman Shades",                                       order={4},                                feature={[4]="Bottom-Up"}},
    all   ={name="Features common to all shades",                                    order={10,01,12},                         feature={[10]="Intermediate stop", [01]="Sync", [12]="Test"}},
}  

```

I think we'd need this data to correctly map a given shade type to all the features that shade supports, taking into account that the same feature id has different semantics for different blind types.

The thing I also need to grok is the `xover` value. the comment says:

```
-- xover (optional - if shade has more than 1 feature) - The crossover point as a percentage value. Below this value, the second feature is used.
```

**Comment by efdeejay** [comment]
> _For more sophisticated top-down / bottom-up support, I was thinking about updating the plugin to provide multiple "services" for a single blind, which HomeKit allows. That way, the UI would provide sliders for both "top-down" and a "bottom-up" services and the plugin could translate those correct feature IDs in the platinum hub (i.e., 4 and 18, respectively). I created a test accessory using the HomeKit Accessory Simulator, and the Home app provides nice side-by-side sliders for each service._

Oh my, if this would be possible, that would be amazing! I know this thread is kind of old by now, but that would be the feature I'm waiting for! :)
